src/backend/mod.rs

```diff
@@ -0,0 +1,1244 @@
+//! Backend: IR to Wasm.
+
+use crate::cfg::CFGInfo;
+use crate::entity::EntityRef;
+use crate::ir::{ExportKind, FuncDecl, FunctionBody, ImportKind, Module, Type, Value, ValueDef};
+use crate::Operator;
+use anyhow::Result;
+use rayon::prelude::*;
+use std::borrow::Cow;
+
+pub mod reducify;
+use reducify::Reducifier;
+pub mod stackify;
+use stackify::{Context as StackifyContext, WasmBlock};
+pub mod treeify;
+use treeify::Trees;
+pub mod localify;
+use localify::Localifier;
+
+pub struct WasmFuncBackend<'a> {
+    body: Cow<'a, FunctionBody>,
+    cfg: CFGInfo,
+}
+
+struct CompileContext<'a> {
+    trees: Trees,
+    ctrl: Vec<WasmBlock<'a>>,
+    locals: Localifier,
+}
+
+macro_rules! op {
+    ($name:tt) => {
+        Some(wasm_encoder::Instruction::$name)
+    };
+}
+
+impl<'a> WasmFuncBackend<'a> {
+    pub fn new(body: &'a FunctionBody) -> Result<WasmFuncBackend<'a>> {
+        body.validate()?;
+        log::debug!("Backend compiling:\n{}\n", body.display_verbose("| ", None));
+        let body = Reducifier::new(body).run();
```

Comment:
Why is the reducifier run in the `new` function here, instead of in the `compile` function? This would be a good to have a comment for, as otherwise it's not obvious that just calling `new` does any meaningful work.

Reply:
Ah, yeah, this one had to do with avoiding a self-referential struct in the `Owned` case of the `Cow` that the reducifier returns; nevertheless the surface API is confusing so I've turned `new` into `compile`, and made it invoke the whole pipeline including old `compile` (now `lower`).